### PR TITLE
Initial draft of TypeScript declaration

### PR DESCRIPTION
**Pull request by jeffrose**
Initial draft of the TypeScript declaration based on discussions in #25. I incorporated some of the changes proposed by @IMax153.

### CODE REVIEW

**Comment by IMax153** [comment]
@jeffrose Really great work on this draft! Looking forward to getting these type definitions out for everyone to use!

I did notice a few things when experimenting with these type definitions. I listed them below for further discussion - curious to hear your thoughts.

1. The `Parser` constructor function is not showing up as an export when I import the module. I believe this is because the `Parser` is not being referenced by the module `export` (i.e. `Arcsecond.ArcsecondStatic`) in the type declarations. This can probably be fixed by adding the `Parser` to the `ArcsecondStatic` interface.

Import_Parser 

2. I am not sure if @francisrstokes wants to expose any of the other types to users of the module (e.g. `ParserState`), but currently they are inaccessible to end users.

3. The `p` method is not (and can not) be marked as private within an interface, so it is currently showing up as an accessible method on constructed parsers. Again I am not sure if this is the intended API or not. 

Parser_p

4. The `ParserState` interface should probably have nullable fields for `result`, `error`, and `data`, given that they are all defaulted to `null` in the initial call to `createParserState`. At the moment, these fields are either their listed type or `undefined`. See below for what I propose:

```typescript
export type ParserState<TResult, TData = null> = {
    isError: boolean;
    error: string | null;
    target: string;
    data: TData;
    index: number;
    result: TResult | null;
};

const createParserState = (target, data = null) => ({
    isError: false,
    error: null,
    target,
    data,
    index: 0,
    result: null
});
```

**Comment by jeffrose** [comment]
Thanks @IMax153! I've pushed some updates based on your feedback. I'm not sure about exposing some of the internal interfaces but I have a spot where I can put them if @francisrstokes thinks it would be useful.

It just occurred to me I've been following the naming convention of starting `type` names with **T** but not the corresponding convention of starting `interface` names with **I**. I should probably be more consistent.

**Comment by IMax153** [comment]
Thanks for incorporating the feedback @jeffrose! In my opinion, it's okay to leave the interfaces without the **I** prefix. A lot of libs out there that expose internal interfaces will leave them un-prefixed to avoid the additional verbosity. Though I do think it makes sense to prefix the generics as you have done.

One small thing is that `data` parameter in `ErrorMapCallback` is missing a type annotation, so it is defaulting to `any`.

As I continue to play around with the typings, I am running a few other issues. I included my thoughts below, but I would be interested to hear what you think about the best way to address them!

Because all parsers under the `ArcsecondParsers` interface are defined as `StringParser`s, Typescript is receiving the incorrect the `TResult` for some of the parsers. Below I included an example, where I supply a `str` parser to the `many` parser. The `many` parser should have a `TResult` of `string[]` but instead infers type `string`. I think we could possibly remedy this issue using the following strategy:
  - Revert back to your original strategy of using `Parser` as the return type for the parsers, and defaulting the `TResult` to `string`. 
  - Define the `TResult` for all functions that return `Parser`s if the returned `Parser` has a `TResult` other than `string` (example below)
  - We should be able to extend the `Parser` interface later to support other return types 

```typescript
// When defined this way, the many parser's TResult is type `string`
many(parser: StringParser): StringParser;
```
Incorrect Result Type

```typescript
// After, converting back to using Parser and supplying the correct TResult
many(parser: Parser): Parser<string[]>;
```
Correct Result Type

**Comment by jeffrose** [comment]
My query-replace of `Parser` to `StringParser` may have been a machete more than a scalpel. Really what I was trying to do was make it less verbose but maybe there are places where `TResult` shouldn't be `string` even in a `StringParser`.

One issue I've noticed is `ParserState.result` is used as a function internally in one place.

https://github.com/francisrstokes/arcsecond/blob/master/index.mjs#L107

Does that mean `ap` shouldn't take `Parser<TResult,TData>` but only takes `Parser<function,TData>`?

I will make some more tweaks and submit an update. @IMax153, do you have a page of use cases you're testing against?

**Comment by IMax153** [comment]
You are right! The `.ap` method seems to expect a `result` that is actually a function. I think the (simplified) type signature would look something like this, given that the applied parser seems to receive the `result` of the initial parser as a parameter:

```typescript
interface Parser<A> {
      ap<B>(parserOfFunction: Parser<(a: A) => B>): Parser<B>;
}
``` 

A few other things I realized when testing the `.ap` method based on your query:
- We are missing types for the `succeedWith` parser
- The parameter supplied to `setData` does not have a type signature 
- The `Applicative`, `Apply`, `Chain`, and `Functor` interfaces seem to add an additional layer of complexity that make it somewhat difficult to resolve certain types. See below for an example of what I mean. This example comes straight from the README on the `.ap` method, but adds in the applied function. Would it just be easier to define the type signatures of each of these methods directly on the `Parser` interface?

Applicative Example

At the moment, I don't have a specific page of use cases I have been testing with. I have just been experimenting with the types as we iterate on them.

I could write up a page of use cases that we could test against and share them if that would be useful? Just let me know!

**Comment by francisrstokes** [comment]
This is awesome work so far guys!

> I'm not sure about exposing some of the internal interfaces but I have a spot where I can put them if @francisrstokes thinks it would be useful.

For 3rd party libraries (like arcsecond-binary) this is probably quite handy.

> I could write up a page of use cases that we could test against and share them if that would be useful? Just let me know!

That's probably a great approach to covering the bases @IMax153.

**Comment by IMax153** [comment]
Alright, I will probably start by making sure that the examples from the README are correctly typed with the declarations and then expand from there. Is there an optimal way for me to share the file with you guys?

Also @jeffrose, see below for what I was thinking in terms of the `Parser` methods. I also made some fixes to the types that were being returned. I also tested the fix to the `.ap` method below and it seems to be working. What do you think?

```typescript
  type ParserState<TResult, TData = void> = {
      isError: boolean;
      error?: string | null;
      target?: string;
      data: TData | null;
      index: number;
      result?: TResult | null;
  };

  type SuccessState<TResult, TData> = ParserState<TResult, TData> & { isError: false };
  type ErrorState<TResult, TData> = ParserState<TResult, TData> & { isError: true };

  type StateData<TResult, TData> = {
      result: TResult;
      data: TData;
  };

  type ErrorStateData<TData> = {
      data: TData;
      error: string;
      index: number;
  }

  type ChainCallback<TInput, TOutput> = (result: TInput) => Parser<TOutput>;
  type ChainFromDataCallback<TInput, TOutput, TData> = (state: StateData<TInput, TData>) => Parser<TOutput, TData>;
  type ErrorChainCallback<TInput, TOutput, TData> = (state: ErrorStateData<TData>) => Parser<TOutput, TData>;
  type ErrorMapCallback<TData> = (error: string, index: number, data: TData) => string;
  type ForkErrorCallback<TResult, TErrorResult, TData> = (error: string, newState: ErrorState<TResult, TData>) => TErrorResult;
  type ForkSuccessCallback<TResult, TSuccessResult, TData> = (result: TResult, newState: SuccessState<TResult, TData>) => TSuccessResult;
  type MapCallback<TInput, TOutput> = (result: TInput) => TOutput;
  type MapDataCallback<TInput, TOutput> = (data: TInput) => TOutput;
  type MapFromDataCallback<TInput, TData, TOutput> = (state: StateData<TInput, TData>) => Parser<TOutput, TData>;

  interface Parser<TResult, TData = void> {
      ['fantasy-land/ap']<TOutput>(parserOfFunction: Parser<(a: TResult) => TOutput, TData>): Parser<TOutput, TData>;
      ['fantasy-land/chain']<TOutput>(fn: ChainCallback<TResult, TOutput>): Parser<TOutput, TData>;
      ['fantasy-land/map']<TOutput>(fn: MapCallback<TResult, TData>): Parser<TOutput, TData>;
      ap<TOutput>(parserOfFunction: Parser<(a: TResult) => TOutput, TData>): Parser<TOutput, TData>;
      chain<TOutput>(fn: ChainCallback<TResult, TOutput>): Parser<TOutput, TData>;
      chainFromData<TOutput>(fn: ChainFromDataCallback<TResult, TOutput, TData>): Parser<TOutput, TData>;
      errorChain<TOutput>(fn: ErrorChainCallback<TResult, TOutput, TData>): Parser<TOutput, TData>;
      errorMap<TData>(fn: ErrorMapCallback<TData>): Parser<TResult, TData>;
      fork<TErrorResult, TSuccessResult>(targetString: string, errorFn: ForkErrorCallback<TResult, TErrorResult, TData>, successFn: ForkSuccessCallback<TResult, TSuccessResult, TData>): SuccessState<TSuccessResult, TData> | ErrorState<TErrorResult, TData>;
      map<TOutput>(fn: MapCallback<TResult, TData>): Parser<TOutput, TData>;
      mapData<TOutput>(fn: MapDataCallback<TData, TOutput>): Parser<TResult, TOutput>;
      mapFromData<TOutput>(fn: MapFromDataCallback<TResult, TData, TOutput>): Parser<TOutput, TData>;
      run(targetString: string): ParserState<TResult, TData>;
  }

  interface ParserConstructor<TResult, TData = void> {
      ['fantasy-land/of']<TResult>(value: TResult): Parser<TResult>;
      of<TResult>(value: TResult): Parser<TResult>;
      new(p: <TInput>(state: ParserState<TInput, TData>) => ParserState<TResult, TData>): Parser<TResult, TData>;
      readonly prototype: Parser<TResult, TData>;
  }
```

**Comment by IMax153** [comment]
Since I had some time over the holiday, I also worked on the declarations for the individual parsers. The signatures and typings can certainly be improved further, and some of the types still aren't working perfectly (i.e. `coroutine`), but I hope that you find this useful as we iterate on the types! 

I also finished a very basic file that we can use to test the types against. Once the interfaces are made public, I can annotate the variables in my file with the return types expected from the parsers. Also let me know how I can share the file with you guys!

```typescript
  type GetDataParser<T> = T extends Parser<any, infer D> ? Parser<D, D> : never;
  type CoroutineNext<T> = T extends Parser<infer TResult> ? TResult : never;
  type NamedParserTuple<K extends string, P extends Parser<any, any>> = [K, P];
  type Either<TResult> = { isError: boolean; value: TResult; }

  interface ArcsecondParsers {
      digit: Parser;
      digits: Parser;
      getData: GetDataParser<Parser>;
      endOfInput: Parser<null>;
      letter: Parser;
      letters: Parser;
      optionalWhitespace: Parser;
      whitespace: Parser;
      anyOfString(s: string): Parser;
      anythingExcept(parser: Parser): Parser;
      between(leftParser: Parser): (rightParser: Parser) => (parser: Parser) => Parser;
      char(c: string): Parser;
      choice(parsers: Iterable<Parser>): Parser;
      composeParsers(parsers: Iterable<Parser>): Parser;
      coroutine<P extends Parser<any, any>, TResult>(g: () => Generator<P, TResult, CoroutineNext<P>>): Parser<TResult>;
      decide<TResult>(fn: (result: TResult) => Parser<TResult>): Parser<TResult>;
      either<TResult>(parser: Parser<TResult>): Parser<Either<TResult>>;
      errorMapTo<TResult = string, TData = null>(fn: (message: string, index: number, data: TData) => string): Parser<TResult, TData>;
      everythingUntil(parser: Parser): Parser;
      fail(errorMessage: string): Parser;
      lookAhead(parser: Parser): Parser;
      many<TResult>(parser: Parser<TResult>): Parser<TResult[]>;
      many1<TResult>(parser: Parser<TResult>): Parser<TResult[]>;
      mapTo<TInput, TOutput>(fn: (data: TInput) => TOutput): Parser<TOutput>;
      mapData<TData, TResult>(fn: (data: TData) => TData): Parser<TResult, TData>;
      namedSequenceOf<K extends string[], P extends Parser<any, any>[]>(pairedParsers: NamedParserTuple<K[number], P[number]>[]): Parser<{ [A in K[number]]: P[number] }>;
      parse<TResult, TData>(parser: Parser<TResult, TData>): (target: string) => TResult;
      pipeParsers(parsers: Iterable<Parser>): Parser;
      possibly<TResult>(parser: Parser<TResult>): Parser<TResult | null>;
      recursiveParser(parserThunk: () => Parser): Parser;
      regex(re: RegExp): Parser;
      sepBy(parser: Parser): (parser: Parser) => Parser;
      sepBy1(parser: Parser): (parser: Parser) => Parser;
      sequenceOf(parsers: Iterable<Parser>): Parser;
      setData<TData, TResult = string>(data: TData): Parser<TResult, TData>;
      skip(parser: Parser): Parser;
      str(s: string): Parser;
      succeedWith<TResult>(value: TResult): Parser<TResult>;
      takeLeft(parser: Parser): (parser: Parser) => Parser;
      takeRight(parser: Parser): (parser: Parser) => Parser;
      tapParser<TData = null, TResult = string>(fn: (state: ParserState<TResult, TData>) => void): Parser<TResult, TData>;
      toPromise<TResult, TData>(result: ParserState<TResult, TData>): Promise<TResult>;
      toValue<TResult, TData>(result: ParserState<TResult, TData>): TResult;
      withData<TData, TResult = string>(parser: Parser<TResult, TData>): (stateData: TData) => Parser<TResult, TData>;
  }
```

**Comment by jeffrose** [comment]
Hey @IMax153, thanks for continuing to work on it. I've had some family and holiday obligations but I will take a look at one you've done and incorporate additional changes into the declaration.

**Comment by jeffrose** [comment]
@IMax153 Great work on the updates! I've incorporated additional changes.

Happy New Year!

**Comment by IMax153** [comment]
Hey @jeffrose! Happy New Year! I hope that you had a wonderful holiday season!

I finally finished up test file for the type declarations. After applying the types to the type tests, I realized that we still have several issues with the type declarations. 

1. The current structure of the declaration file prevents us from properly exporting certain types and keeping other types private.
2. The interfaces that should be public for users that wish to extend the library are currently still private.
3. I realized that I made several errors in the typings that I most recently sent to you.

As I result of this I spent some (probably too much) time over the past few days revamping the declarations entirely and restructuring the type declarations file. I followed a similar pattern that the React devs used for their [type declarations](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts).

Rather than pasting my type testing file and updated declarations here, I pushed a [new branch](https://github.com/IMax153/arcsecond/blob/feature/types/index.d.ts) to my local fork of the repo so you can take a look and let me know what you think. I annotated the types that are still problematic with `TODO:` declarations above the JSDoc.

**Comment by jeffrose** [comment]
I think your revamp looks good. Unfortunately with the start of the new year, my job is ramping up and I do not have much time to work on this right now. @francisrstokes may want to reject my PR and start with yours instead.

**Comment by IMax153** [comment]
@jeffrose No problem! I totally understand, I appreciate you letting me know! We can figure all this out when @francisrstokes gets back from his trip to Southeast Asia.

**Comment by francisrstokes** [comment]
Hey Jeff - thanks for all the work you've put in pushing this through. Whatever ends up happening with the PR I'll make sure you end up in the contributions list when it's merged!
Good luck in the job!

**Comment by IMax153** [comment]
Agreed - @jeffrose, I have greatly appreciated collaborating on this with and look forward to continuing to do so in whatever capacity you are able to moving forward! Best of luck with the new job!

@francisrstokes how would you like to proceed moving forward? I can submit a PR to work off of with the declarations that Jeff and I have been working on if we want to go that direction?

**Comment by francisrstokes** [comment]
Hey @IMax153 - again, apologies for getting back so late - I just arrived back home and this is the first time sitting at an actual computer looking at github (can't say that I didn't miss it! 😄)

Your approach sounds fine by me! The direction this effort has been going in is great, I don't see any reason to change anything drastically. Thanks for the continued interest, it makes a massive difference!

**Comment by IMax153** [comment]
Hey @francisrstokes! No worries at all! I hope that you enjoyed your trip! I will create my own PR to replace this one in the coming days. Just trying to address a few more issues with the type declarations first. It will be good to have another set of eyes on them again!

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    type StateData<TResult, TData> = {
+        result: TResult;
+        data: TData;
+    };
+
+    type ChainCallback<TInput, TOutput> = (result: TInput) => Parser<TOutput>;
+    type ChainFromDataCallback<TInput, TOutput, TData> = (state: StateData<TInput, TData>) => Parser<TOutput, TData>;
+    type ErrorChainCallback<TInput, TOutput, TData> = (state: ErrorState<TInput, TData>) => Parser<TOutput, TData>;
+    type ErrorMapCallback = (error: string, index: number, data) => string;
```
Comment (gregolai) [inline index.d.ts:26]: `type ErrorMapCallback<TData> = (error: string, index: number, data: TData) => string;`

and in the Parser interface:

`errorMap(fn: ErrorMapCallback<TData>): Parser<TResult, TData>;`

**Pull request closed without merging**